src/operation_dispatcher.rs

```diff
@@ -0,0 +1,291 @@
+impl Operation {
+    pub fn new(extension: Rc<Extension>, procedure: Procedure) -> Self {
+        Self {
+            state: State::Pending,
+            result: Err(Status::Empty),
+            extension,
+            procedure,
+            grpc_call,
+            get_map_values_bytes,
+        }
+    }
+
+    fn trigger(&mut self) {
+        if let State::Done = self.state {
+        } else {
+            self.result = self.procedure.0.send(
+                self.get_map_values_bytes,
+                self.grpc_call,
+                self.procedure.1.clone(),
+            );
+            self.state.next();
+        }
+    }
+
+    pub fn get_state(&self) -> State {
+        self.state.clone()
+    }
+
+    pub fn get_result(&self) -> Result<u32, Status> {
+        self.result
+    }
+
+    pub fn get_extension_type(&self) -> ExtensionType {
+        self.extension.extension_type.clone()
+    }
+
+    pub fn get_failure_mode(&self) -> FailureMode {
+        self.extension.failure_mode.clone()
+    }
+}
+
+#[allow(dead_code)]
+pub struct OperationDispatcher {
+    operations: RefCell<Vec<Operation>>,
+    service_handlers: HashMap<String, Rc<GrpcServiceHandler>>,
+}
+
+#[allow(dead_code)]
+impl OperationDispatcher {
+    pub fn default() -> Self {
+        OperationDispatcher {
+            operations: RefCell::new(vec![]),
+            service_handlers: HashMap::default(),
+        }
+    }
+    pub fn new(service_handlers: HashMap<String, Rc<GrpcServiceHandler>>) -> Self {
+        Self {
+            service_handlers,
+            operations: RefCell::new(vec![]),
+        }
+    }
+
+    pub fn build_operations(
+        &self,
+        policy: &Policy,
+        descriptors: RepeatedField<RateLimitDescriptor>,
+    ) {
+        let mut operations: Vec<Operation> = vec![];
+        policy.actions.iter().for_each(|action| {
+            // TODO(didierofrivia): Error handling
+            if let Some(service) = self.service_handlers.get(&action.extension) {
+                let message = GrpcMessage::new(
+                    service.get_extension_type(),
+                    policy.domain.clone(),
+                    descriptors.clone(),
+                );
+                operations.push(Operation::new(
+                    service.get_extension(),
+                    (Rc::clone(service), message),
+                ))
+            }
+        });
+        self.push_operations(operations);
+    }
+
+    pub fn push_operations(&self, operations: Vec<Operation>) {
+        self.operations.borrow_mut().extend(operations);
+    }
+
+    pub fn get_current_operation_state(&self) -> Option<State> {
+        self.operations
+            .borrow()
+            .first()
+            .map(|operation| operation.get_state().clone())
+    }
+
+    pub fn get_current_operation_result(&self) -> Result<u32, Status> {
+        self.operations.borrow().first().unwrap().get_result()
+    }
+
+    pub fn next(&self) -> Option<Operation> {
```

Comment:
I think this signature will need to be somewhat rethought... Probably something along the lines of a 

 - `.start()` that starts a "sequence", `initial` state
 - `.next(...)` with some params (gRPC response's `token_id` & `response_size`?) to go from one operation to the next
 - `()` i.e. "void"/unit, i.e. nothing left to do, representing the `terminated` state

Reply:
Yup, after discussing about how would it be consumed `on_grpc_response` and the requirements to digest the actual response needing the `response_size` makes sense. Will shoot a following PR addressing this. Thanks!